File: crates/dsp-runtime/src/protocol/recording.rs
use serde::{Deserialize, Serialize};

use super::BinaryPayload;

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct RecordingStartConfig {
    pub path: String,
    pub asset_id: String,
    pub originator: String,
    pub origination_date: String,
    pub origination_time: String,
    pub time_reference: i64,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct RecordingResult {
    pub path: String,
    pub sample_rate: u32,
    pub channels: u32,
    pub frame_count: i64,
    pub dropout_frames: i64,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct RecordingWaveform {
    pub sample_rate: u32,
    pub channels: u32,
    pub frame_count: i64,
    pub start_frame: i64,
    pub end_frame: i64,
    pub frames_per_bucket: u32,
    pub bucket_count: u32,
    pub peaks: BinaryPayload,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct MidiRecordingTakeConfig {
    pub path: String,
    pub source_id: String,
    pub clip_id: String,
    pub track_id: String,
    /// Stable MIDI input port id, or `None` for all inputs.
    pub port_id: Option<String>,
    pub channel: Option<u8>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct MidiRecordingStartConfig {
    pub takes: Vec<MidiRecordingTakeConfig>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct MidiRecordingPreviewNote {
    pub id: u32,
    pub start_tick: u64,
    pub end_tick: u64,
    pub channel: u8,
    pub key: u8,
    pub velocity: u8,
    pub active: bool,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct MidiRecordingTakePreview {
    pub clip_id: String,
    pub track_id: String,
    pub notes: Vec<MidiRecordingPreviewNote>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct MidiRecordingPreview {
    pub position_tick: u64,
    pub takes: Vec<MidiRecordingTakePreview>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct MidiRecordingTakeResult {
    pub path: String,
    pub source_id: String,
    pub clip_id: String,
    pub track_id: String,
    pub event_count: u64,
    pub dropped_events: u64,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct MidiRecordingResult {
    pub takes: Vec<MidiRecordingTakeResult>,
}

File: apps/desktop/src/main/audio-host/wire/recording.ts
import type { BinaryPayloadWire } from "./binary"

export interface AudioHostRecordingConfig {
  path: string
  assetId: string
  originator: string
  originationDate: string
  originationTime: string
  timeReference: number
  sampleRate: number
  channels: number
}

export interface AudioHostRecordingResultWire {
  path: string
  sample_rate: number
  channels: number
  frame_count: number
  dropout_frames: number
}

export interface AudioHostMidiRecordingTakeConfig {
  path: string
  sourceId: string
  clipId: string
  trackId: string
  portId: string | null
  channel: number | null
}

export interface AudioHostMidiRecordingConfig {
  takes: AudioHostMidiRecordingTakeConfig[]
}

export interface AudioHostMidiRecordingTakeResultWire {
  path: string
  source_id: string
  clip_id: string
  track_id: string
  event_count: number
  dropped_events: number
}

export interface AudioHostMidiRecordingResultWire {
  takes: AudioHostMidiRecordingTakeResultWire[]
}

export interface AudioHostMidiRecordingTakeResult {
  path: string
  sourceId: string
  clipId: string
  trackId: string
  eventCount: number
  droppedEvents: number
}

export interface AudioHostMidiRecordingResult {
  takes: AudioHostMidiRecordingTakeResult[]
}

export interface AudioHostWaveformWire {
  sample_rate: number
  channels: number
  frame_count: number
  start_frame: number
  end_frame: number
  frames_per_bucket: number
  bucket_count: number
  peaks: BinaryPayloadWire
}

export interface AudioHostRecordingResult {
  path: string
  sampleRate: number
  channels: number
  frameCount: number
  dropoutFrames: number
}

export interface AudioHostWaveform {
  sampleRate: number
  channels: number
  frameCount: number
  startFrame: number
  endFrame: number
  framesPerBucket: number
  bucketCount: number
  peaks: Uint8Array
}
